File: packages/learning/src/embeddings.ts
/**
 * Embedding local por "hashing trick" (feature hashing) — vetor de
 * frequência de termos, hasheado para uma dimensão fixa e normalizado
 * (L2). Não é uma rede neural, mas é matemática real de recuperação por
 * similaridade (mesma família de PLN clássico usada em produção, ex.:
 * Vowpal Wabbit): funciona sem chamar nenhum provedor externo, então a
 * memória indexada funciona mesmo sem um provedor LLM configurado. Se um
 * dia trocar por embeddings de verdade (API), só esta função muda — o
 * resto do pacote não sabe a diferença.
 */

const DIMENSIONS = 128;

function tokenize(text: string): string[] {
  return (
    text
      .normalize("NFD")
      .replace(/[̀-ͯ]/g, "")
      .toLowerCase()
      .match(/[a-z0-9]+/g) ?? []
  );
}

function hashToken(token: string): number {
  let hash = 0;
  for (let i = 0; i < token.length; i++) {
    hash = (hash * 31 + token.charCodeAt(i)) >>> 0;
  }
  return hash % DIMENSIONS;
}

export function embedText(text: string): number[] {
  const vector = new Array(DIMENSIONS).fill(0);
  for (const token of tokenize(text)) {
    vector[hashToken(token)] += 1;
  }
  const norm = Math.sqrt(vector.reduce((sum, v) => sum + v * v, 0)) || 1;
  return vector.map((v) => v / norm);
}

File: packages/learning/hermes_plugin/embeddings.py
"""Embedding local por *hashing trick* — porte fiel de `src/embeddings.ts`.

Mesmo vetor, mesma dimensão, mesma normalização: um índice gerado pelo lado
TypeScript pode ser consultado pelo Python e vice-versa. `tests/test_parity.py`
compara as duas implementações contra uma fixture gerada pelo TypeScript — se
alguém mexer num lado só, o teste quebra.

Não é rede neural: é recuperação por similaridade de PLN clássico (mesma
família do Vowpal Wabbit). A vantagem é não depender de nenhuma API externa,
que é o que permite a memória funcionar sem provedor LLM configurado.
"""
from __future__ import annotations

import math
import re
import unicodedata

DIMENSIONS = 128

_TOKEN = re.compile(r"[a-z0-9]+")


def tokenize(text: str) -> list[str]:
    normalized = unicodedata.normalize("NFD", text)
    without_marks = "".join(char for char in normalized if not unicodedata.combining(char))
    return _TOKEN.findall(without_marks.lower())


def hash_token(token: str) -> int:
    # `(hash * 31 + charCode) >>> 0` do TypeScript: inteiro sem sinal de 32 bits.
    digest = 0
    for char in token:
        digest = (digest * 31 + ord(char)) & 0xFFFFFFFF
    return digest % DIMENSIONS


def embed_text(text: str) -> list[float]:
    vector = [0.0] * DIMENSIONS
    for token in tokenize(text):
        vector[hash_token(token)] += 1.0
    norm = math.sqrt(sum(value * value for value in vector)) or 1.0
    return [value / norm for value in vector]


def cosine(left: list[float], right: list[float]) -> float:
    """Vetores já vêm normalizados (L2), então o produto interno é o cosseno."""
    return sum(a * b for a, b in zip(left, right))
